### PR TITLE
docs: add release note for 0.4.0

### PR DESCRIPTION
**Pull request by hardfist**

### CODE REVIEW

File: docs/en/blog/announcing-0.4.mdx [hunk 1]
```diff
@@ -0,0 +1,262 @@
+# Rspack 0.4 Release Announcement
+
+## Announcing Rsbuild v0.1
+
+![](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/a0fe4dcdc2264c6f81a72ddeabd51268~tplv-k3u1fbpfcp-jj-mark:0:0:0:0:q75.image#?w=1800&h=579&s=145010&e=png&b=fffdf1)
```
Comment (hardfist) [inline docs/en/blog/announcing-0.4.mdx:5]: should use local image

File: docs/zh/blog/announcing-0.4.mdx [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+1. 移除对一些 [[内置](https://www.rspack.dev/config/builtins)](https://www.rspack.dev/config/builtins) 功能的支持：
+
+- [[builtins.relay](https://www.rspack.dev/config/builtins#builtinsrelay)](https://www.rspack.dev/config/builtins#builtinsrelay)：移动到 `rspackExperiments.relay`
+- [[builtins.react](https://www.rspack.dev/config/builtins#builtinsreact)](https://www.rspack.dev/config/builtins#builtinsreact)：移动到 `jsc.transform.react`
+- [[builtins.emotion](https://www.rspack.dev/config/builtins#builtinsemotion)](https://www.rspack.dev/config/builtins#builtinsemotion)：移动到 `rspackExperiments.emotion`
+- [[builtins.pluginImport](https://www.rspack.dev/config/builtins#builtinspluginimport)](https://www.rspack.dev/config/builtins#builtinspluginimport)：移动到 `rspackExperiments.import`
+- [[builtins.decorator](https://www.rspack.dev/config/builtins#builtinsdecorator)](https://www.rspack.dev/config/builtins#builtinsdecorator)：移动到 `jsc.parser.decorator`
+- [[builtins.presetEnv](https://www.rspack.dev/config/builtins#builtinspresetenv)](https://www.rspack.dev/config/builtins#builtinspresetenv)：移动到 `jsc.env`
```
Comment (ahabhgk) [inline docs/zh/blog/announcing-0.4.mdx:58]:
```suggestion
1. 移除对一些[内置](https://www.rspack.dev/config/builtins)功能的支持：

- [builtins.relay](https://www.rspack.dev/config/builtins#builtinsrelay)：移动到 `rspackExperiments.relay`
- [builtins.react](https://www.rspack.dev/config/builtins#builtinsreact)：移动到 `jsc.transform.react`
- [builtins.emotion](https://www.rspack.dev/config/builtins#builtinsemotion)：移动到 `rspackExperiments.emotion`
- [builtins.pluginImport](https://www.rspack.dev/config/builtins#builtinspluginimport)：移动到 `rspackExperiments.import`
- [builtins.decorator](https://www.rspack.dev/config/builtins#builtinsdecorator)：移动到 `jsc.parser.decorator`
- [builtins.presetEnv](https://www.rspack.dev/config/builtins#builtinspresetenv)：移动到 `jsc.env`
```

File: docs/zh/blog/announcing-0.4.mdx [hunk 3]
```diff
@@ -0,0 +1,247 @@
+# Rspack 0.4 发布公告
+
+## Rsbuild v0.1 发布公告
+
+![](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/a0fe4dcdc2264c6f81a72ddeabd51268~tplv-k3u1fbpfcp-jj-mark:0:0:0:0:q75.image#?w=1800&h=579&s=145010&e=png&b=fffdf1)
+
+我们很高兴地宣布 **[Rsbuild](https://github.com/web-infra-dev/rsbuild)** **v0.1** 的发布！这个里程碑表明 Rsbuild API 现在更加稳定可靠。
+
+Rsbuild 是基于 Rspack 的构建工具，旨在成为 **更加用户友好和强大的增强型 Rspack CLI**。由于 Rspack 团队对 Web 构建最佳实践的探索和实施，Rsbuild 是那些希望从 Webpack 迁移至 Rspack 的理想解决方案。它能够显著减少 90% 的配置同时提供 10 倍的构建速度。
```
Comment (bvanjoi) [inline docs/zh/blog/announcing-0.4.mdx:9]: 5 ~ 10

File: docs/zh/blog/announcing-0.4.mdx [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  },
+  experiments: {
+    rspackFuture: {
+      disableTransformByDefault: true
+    }
+  }
+};
+```
+
+2. [[target](https://www.rspack.dev/config/target)](https://www.rspack.dev/config/target) 不会降级用户端代码
```
Comment (ahabhgk) [inline docs/zh/blog/announcing-0.4.mdx:95]:
```suggestion
2. [target](https://www.rspack.dev/config/target) 不再降级用户端代码
```

File: docs/zh/blog/announcing-0.4.mdx [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
++        }
+        }
+        type: 'javascript/auto',
+      },
+    ],
+  }
+};
+```
+
+3. 移除非 webpack 兼容的 [[Rule.type](```
```
Comment (ahabhgk) [inline docs/zh/blog/announcing-0.4.mdx:124]: [[Rule.type](``` ??

File: docs/zh/blog/announcing-0.4.mdx [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+## 废弃 builtin.react.refresh
+
+由于在 v0.4.0 中默认启用了 `experiments.rspackFuture.disableTransformByDefault`，`builtin.react.refresh` 也已被废弃。现在我们建议使用 `@rspack/plugin-react-refresh` 来启用 React 快速刷新。
```
Comment (ahabhgk) [inline docs/zh/blog/announcing-0.4.mdx:128]:
```suggestion
由于在 v0.4.0 中默认启用了 `experiments.rspackFuture.disableTransformByDefault`，`builtin.react.refresh` 也已被废弃。现在我们建议使用 `@rspack/plugin-react-refresh` 来启用 React Fast Refresh。你需要手动安装 [@rspack/plugin-react-refresh](https://www.npmjs.com/package/@rspack/plugin-react-refresh) 和 [react-refresh](https://www.npmjs.com/package/react-refresh)。
```

File: docs/zh/blog/announcing-0.4.mdx [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+### 废弃默认转换
+
+`experiments.rspackFuture.disableTransformByDefault` 在 v0.4.0 中默认启用。对于仍需要旧行为的用户，可以手动将此选项设置为 `false`。
+
+此功能主要解决三类问题：[内置](https://www.rspack.dev/config/builtins) 代码转换功能，[目标](https://www.rspack.dev/config/target)，和自定义 [Rule.type](https://www.rspack.dev/config/module#ruletype)。
+
+1. 移除对一些 [[内置](https://www.rspack.dev/config/builtins)](https://www.rspack.dev/config/builtins) 功能的支持：
+
+- [[builtins.relay](https://www.rspack.dev/config/builtins#builtinsrelay)](https://www.rspack.dev/config/builtins#builtinsrelay)：移动到 `rspackExperiments.relay`
```
Comment (bvanjoi) [inline docs/zh/blog/announcing-0.4.mdx:53]:
```
[builtins.relay](https://www.rspack.dev/config/builtins#builtinsrelay)
```

In my opinion, this format is sufficient, there's no need to display its URL. Same as follows.

File: docs/zh/blog/announcing-0.4.mdx [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+查看更多详情，请访问 [这里](https://www.rspack.dev/zh/guide/react.html#fast-refresh)。
+
+### 废弃 builtin:sass-loader
+
+`builtin:sass-loader` 现已被废弃。如果您正在使用它，请迁移到 `sass-loader`。Rspack 将在 v0.5.0 中移除 `builtin:sass-loader`。
+
+### 废弃 experiments.incrementalRebuild
+
+`experiments.incrementalRebuild` 现已被废弃。Rspack 将在 v0.5.0 中移除它。
```
Comment (ahabhgk) [inline docs/zh/blog/announcing-0.4.mdx:180]:
```suggestion
`experiments.incrementalRebuild` 已内置开启，所以不再需要改配置。Rspack 将在 v0.5.0 中移除该配置。
```

File: docs/zh/blog/announcing-0.4.mdx [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+}
+```
+
+我们建议迁移到 `builtin:swc-loader` 来进行模块转换，更多详情请参阅 [Deprecating default transformation](#废弃默认转换)。
+
+### 对于 React 应用程序，请使用 `@rspack/plugin-react-refresh` 来支持 Fast Refresh
+
+当我们禁用默认转换时，`builtin.react.refresh` 将无法工作，因此您需要使用 `@rspack/plugin-react-refresh` 来启用 Fast Refresh，更多细节请参阅 [Deprecating builtin.react.refresh](#废弃-builtinreactrefresh)。
+
+### 迁移 builtin options 到 builtin plugins
```
Comment (bvanjoi) [inline docs/zh/blog/announcing-0.4.mdx:247]: Is it just a heading?

File: docs/zh/blog/announcing-0.4.mdx [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+};
+```
+
+查看更多详情，请访问 [这里](https://www.rspack.dev/zh/guide/react.html#fast-refresh)。
+
+### 废弃 builtin:sass-loader
+
+`builtin:sass-loader` 现已被废弃。如果您正在使用它，请迁移到 `sass-loader`。Rspack 将在 v0.5.0 中移除 `builtin:sass-loader`。
+
+### 废弃 experiments.incrementalRebuild
```
Comment (ahabhgk) [inline docs/zh/blog/announcing-0.4.mdx:178]:
```suggestion
### experiments.incrementalRebuild 配置废弃
```

File: docs/zh/blog/announcing-0.4.mdx [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+
+以前，一些 API 可能会通过从 @rspack/core 重新导出而被意外导出。现在通过此重构，我们从 @rspack/core 中清理了导出的 API。
+
+这不应该造成任何问题，但如果您意外导出了 API，则可能会出现问题，您可能会以不正确的方式使用 Rspack。
+
+如果确实有需要从此重构中移除的 API，请在 Rspack 仓库中提出问题。
+
+### 废弃 `builtins.devFriendlySplitChunks` 和 `experiments.new_split_chunks`
+
+为了完全迁移到 Webpack 的分割块实现，这些字段已被废弃。Rspack 将在 v0.5.0 中移除这些字段。
```
Comment (ahabhgk) [inline docs/zh/blog/announcing-0.4.mdx:192]:
```suggestion
为了完全迁移到 Webpack 的 splitChunks 实现，这些字段已被废弃。Rspack 将在 v0.5.0 中移除这些字段。
```

File: docs/en/blog/announcing-0.4.mdx [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+
+1.  Removal of support for some [[builtins](https://www.rspack.dev/config/builtins)](https://www.rspack.dev/config/builtins) features:
+
+- [[builtins.relay](https://www.rspack.dev/config/builtins#builtinsrelay)](https://www.rspack.dev/config/builtins#builtinsrelay): moved to `rspackExperiments.relay`
+- [[builtins.react](https://www.rspack.dev/config/builtins#builtinsreact)](https://www.rspack.dev/config/builtins#builtinsreact): moved to `jsc.transform.react`
+- [[builtins.emotion](https://www.rspack.dev/config/builtins#builtinsemotion)](https://www.rspack.dev/config/builtins#builtinsemotion): moved to `rspackExperiments.emotion`
+- [[builtins.pluginImport](https://www.rspack.dev/config/builtins#builtinspluginimport)](https://www.rspack.dev/config/builtins#builtinspluginimport): moved to `rspackExperiments.import`
+- [[builtins.decorator](https://www.rspack.dev/config/builtins#builtinsdecorator)](https://www.rspack.dev/config/builtins#builtinsdecorator): moved to `jsc.parser.decorator`
+- [[builtins.presetEnv](https://www.rspack.dev/config/builtins#builtinspresetenv)](https://www.rspack.dev/config/builtins#builtinspresetenv): moved to `jsc.env`
```
Comment (ahabhgk) [inline docs/en/blog/announcing-0.4.mdx:58]:
```suggestion
1.  Removal of support for some [builtins](https://www.rspack.dev/config/builtins) features:

- [builtins.relay](https://www.rspack.dev/config/builtins#builtinsrelay): moved to `rspackExperiments.relay`
- [builtins.react](https://www.rspack.dev/config/builtins#builtinsreact): moved to `jsc.transform.react`
- [builtins.emotion](https://www.rspack.dev/config/builtins#builtinsemotion): moved to `rspackExperiments.emotion`
- [builtins.pluginImport](https://www.rspack.dev/config/builtins#builtinspluginimport): moved to `rspackExperiments.import`
- [builtins.decorator](https://www.rspack.dev/config/builtins#builtinsdecorator): moved to `jsc.parser.decorator`
- [[builtins.presetEnv](https://www.rspack.dev/config/builtins#builtinspresetenv)](https://www.rspack.dev/config/builtins#builtinspresetenv): moved to `jsc.env`
```

File: docs/en/blog/announcing-0.4.mdx [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+  },
+  experiments: {
+    rspackFuture: {
+      disableTransformByDefault: true
+    }
+  }
+};
+```
+
+2.  [[target](https://www.rspack.dev/config/target)](https://www.rspack.dev/config/target) will not downgrade user-side code
```
Comment (ahabhgk) [inline docs/en/blog/announcing-0.4.mdx:96]:
```suggestion
2. [target](https://www.rspack.dev/config/target) will not downgrade user-side code
```

File: docs/en/blog/announcing-0.4.mdx [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+        options: {
+          jsc: {
+            parser: {
+              syntax: "ecmascript"
+            },
++           target: "es5" // Notice: `jsc.target` and `env` cannot be set at the same time.
+          },
++        env: { //  Notice: `jsc.target` and `env` cannot be set at the same time.
++         targets: "chrome >= 48"
++        }
```
Comment (ahabhgk) [inline docs/en/blog/announcing-0.4.mdx:116]:
```suggestion
+         env: { //  Notice: `jsc.target` and `env` cannot be set at the same time.
+           targets: "chrome >= 48"
+         }
```

File: docs/en/blog/announcing-0.4.mdx [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+        }
+        type: 'javascript/auto',
+      },
+    ],
+  }
+};
+```
+
+3.  Removed non-webpack compatible [[Rule.type](https://www.rspack.dev/config/module#ruletype)](https://www.rspack.dev/config/module#ruletype)
```
Comment (ahabhgk) [inline docs/en/blog/announcing-0.4.mdx:125]:
```suggestion
3.  Removed non-webpack compatible [Rule.type](https://www.rspack.dev/config/module#ruletype)
```

File: docs/en/blog/announcing-0.4.mdx [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+- `"javascript/esm"`
+- `"javascript/dynamic"`
+
+Refer to [this](https://www.rspack.dev/config/experiments.html#experimentsrspackfuturedisabletransformbydefault) for the complete migration guide.
+
+Check out our previous discussion [here](https://github.com/web-infra-dev/rspack/discussions/4070).
+
+### Deprecating builtin.react.refresh
+
+With `experiments.rspackFuture.disableTransformByDefault` is enabled by default in v0.4.0, `builtin.react.refresh` has also been deprecated. Now we recommend using `@rspack/plugin-react-refresh` to enable react fast refresh.
```
Comment (ahabhgk) [inline docs/en/blog/announcing-0.4.mdx:145]:
```suggestion
With `experiments.rspackFuture.disableTransformByDefault` is enabled by default in v0.4.0, `builtin.react.refresh` has also been deprecated. Now we recommend using `@rspack/plugin-react-refresh` to enable react fast refresh. And to use it, you need install [@rspack/plugin-react-refresh](https://www.npmjs.com/package/@rspack/plugin-react-refresh) and [react-refresh](https://www.npmjs.com/package/react-refresh)
```

File: docs/en/blog/announcing-0.4.mdx [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+};
+```
+
+Checkout [here](https://www.rspack.dev/zh/guide/react.html#fast-refresh) for more details.
+
+### Deprecating builtin:sass-loader
+
+`builtin:sass-loader` has now been deprecated. If you are using it, migrate to `sass-loader`. Rspack will remove `builtin:sass-loader` in v0.5.0.
+
+### Deprecating experiments.incrementalRebuild
```
Comment (ahabhgk) [inline docs/en/blog/announcing-0.4.mdx:195]:
```suggestion
### Deprecating experiments.incrementalRebuild options
```

File: docs/en/blog/announcing-0.4.mdx [hunk 18]
```diff
@@ -0,0 +1,262 @@
+# Rspack 0.4 Release Announcement
+
+## Announcing Rsbuild v0.1
+
+![](/rsbuild.png)
+
+We are pleased to announce **the release of** **[Rsbuild](https://github.com/web-infra-dev/rsbuild)** **v0.1!** This milestone indicates that the Rsbuild API is now more stable and reliable to use.
+
+Rsbuild is an Rspack-based build tool, designed to be **an enhanced Rspack** **CLI** that is both more user friendly and powerful. As a result of the Rspack team's exploration and implementation of best practices for web building, Rsbuild is the ideal solution for those looking to migrate from Webpack to Rspack. It significantly reduces configuration by 90% while offering a 10x build speed.
```
Comment (bvanjoi) [inline docs/en/blog/announcing-0.4.mdx:9]:
```suggestion
Rsbuild is an Rspack-based build tool, designed to be **an enhanced Rspack** **CLI** that is both more user friendly and powerful. As a result of the Rspack team's exploration and implementation of best practices for web building, Rsbuild is the ideal solution for those looking to migrate from Webpack to Rspack. It significantly reduces configuration by 90% while offering a 5~10x build speed.
```

File: docs/zh/blog/announcing-0.4.mdx [hunk 19]
```diff
@@ -0,0 +1,247 @@
+# Rspack 0.4 发布公告
+
+## Rsbuild v0.1 发布公告
+
+![](/rsbuild.png)
+
+我们很高兴地宣布 **[Rsbuild](https://github.com/web-infra-dev/rsbuild)** **v0.1** 的发布！这个里程碑表明 Rsbuild API 现在更加稳定可靠。
+
+Rsbuild 是基于 Rspack 的构建工具，旨在成为 **更加用户友好和强大的增强型 Rspack CLI**。由于 Rspack 团队对 Web 构建最佳实践的探索和实施，Rsbuild 是那些希望从 Webpack 迁移至 Rspack 的理想解决方案。它能够显著减少 90% 的配置同时提供 10 倍的构建速度。
```
Comment (bvanjoi) [inline docs/zh/blog/announcing-0.4.mdx:9]:
```suggestion
Rsbuild 是基于 Rspack 的构建工具，旨在成为 **更加用户友好和强大的增强型 Rspack CLI**。由于 Rspack 团队对 Web 构建最佳实践的探索和实施，Rsbuild 是那些希望从 Webpack 迁移至 Rspack 的理想解决方案。它能够显著减少 90% 的配置同时提供 5 ～ 10 倍的构建速度。
```

File: docs/en/blog/announcing-0.4.mdx [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+
+Checkout [here](https://www.rspack.dev/zh/guide/react.html#fast-refresh) for more details.
+
+### Deprecating builtin:sass-loader
+
+`builtin:sass-loader` has now been deprecated. If you are using it, migrate to `sass-loader`. Rspack will remove `builtin:sass-loader` in v0.5.0.
+
+### Deprecating experiments.incrementalRebuild
+
+`experiments.incrementalRebuild` has now been deprecated. Rspack will remove it in v0.5.0.
```
Comment (ahabhgk) [inline docs/en/blog/announcing-0.4.mdx:197]:
```suggestion
`experiments.incrementalRebuild` has now internally enabled, so you don't need the options anymore, the options has been deprecated. Rspack will remove the options in v0.5.0.
```

File: docs/zh/blog/announcing-0.4.mdx [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+
+查看更多详情，请访问 [这里](https://www.rspack.dev/zh/guide/react.html#fast-refresh)。
+
+### 废弃 builtin:sass-loader
+
+`builtin:sass-loader` 现已被废弃。如果您正在使用它，请迁移到 `sass-loader`。Rspack 将在 v0.5.0 中移除 `builtin:sass-loader`。
+
+### experiments.incrementalRebuild 配置废弃
+
+`experiments.incrementalRebuild` 已内置开启，所以不再需要改配置。Rspack 将在 v0.5.0 中移除该配置。
```
Comment (ahabhgk) [inline docs/zh/blog/announcing-0.4.mdx:190]:
```suggestion
`experiments.incrementalRebuild` 已内置开启，所以不再需要该配置。Rspack 将在 v0.5.0 中移除该配置。
```

**Review by h-a-n-a — APPROVED** [review]

File: docs/en/blog/announcing-0.4.mdx [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+
+`@rspack/core` is now a peer dependency of `@rspack/cli` rather than a direct dependency. This means that you need to manually install `@rspack/core` with `@rspack/cli` now. aligning Rspack more closely with webpack. In the long term, the positioning of `@rspack/cli` will no longer be an out-of-the-box solution. We will align `@rspack/cli` with webpack-cli and may even directly support the use of `@rspack/core` in `webpack-cli`. We recommend [Rsbuild](https://rsbuild.dev/) as an out-of-the-box solution.
+
+### Deprecating default transformation
+
+`experiments.rspackFuture.disableTransformByDefault` is enabled by default in v0.4.0. For people that still need the legacy behavior, you may manually set this option to `false`.
+
+This feature primarily addresses three categories of problems: [builtins](https://www.rspack.dev/config/builtins) code transformation features, [target](https://www.rspack.dev/config/target), and custom [Rule.type](https://www.rspack.dev/config/module#ruletype).
+
+1. Removal of support for some [[builtins](https://www.rspack.dev/config/builtins) features:
```
Comment (h-a-n-a) [inline docs/en/blog/announcing-0.4.mdx:61]:
```suggestion
1. Removal of support for some [builtins](https://www.rspack.dev/config/builtins) features:
```

**Pull request merged**